### ISSUE TITLE
GarminConnect Login Error

### ISSUE THREAD

**Issue by AngelaKru** [magsol/garmin#10]
Anyone getting the following error? Seemed to start at the weekend for me, things worked find last week but now I can't download my garmin files any longer. I can log into the website no problem - could this be related to the ongoing changes they're making with the API??

> Login successful! Proceeding...
> Traceback (most recent call last):
>   File "_folderlocationhere_\GitHub\garmin-master\garmin-master\download.py", line 246, in <module>
>     download_files_for_user(username, password, output)
>   File "_folderlocationhere_\GitHub\garmin-master\garmin-master\download.py", line 195, in download_files_for_user
>     login(agent, username, password)
>   File "_folderlocationhere_\GitHub\garmin-master\garmin-master\download.py", line 98, in login
>     agent.open(response_url)
>   File "C:\Program Files (x86)\Python27\lib\site-packages\mechanize\_mechanize.py", line 203, in open
>     return self._mech_open(url, data, timeout=timeout)
>   File "C:\Program Files (x86)\Python27\lib\site-packages\mechanize\_mechanize.py", line 230, in _mech_open
>     response = UserAgentBase.open(self, request, data)
>   File "C:\Program Files (x86)\Python27\lib\site-packages\mechanize\_opener.py", line 188, in open
>     req = meth(req)
>   File "C:\Program Files (x86)\Python27\lib\site-packages\mechanize\_urllib2_fork.py", line 1043, in do_request_
>     **raise URLError('no host given')**
> urllib2.URLError: <urlopen error no host given>

**Comment by magsol** [comment]
More than likely; they keep updating how the site works.

From your error logs, it actually looks at though the login itself still worked ok, but rather it was the part where the actual data files are downloaded that it failed. I'm hoping this means it'll be somewhat easier to fix; the URL just may need to be updated to a new version.

**Comment by AngelaKru** [comment]
It's before that point that it's failing

>     # Submit the login!
>     res = agent.submit()
>     if res.get_data().find("Invalid") >= 0:
>         quit("Login failed! Check your credentials, or submit a bug report.")
>     elif res.get_data().find("SUCCESS") >= 0:
>         print 'Login successful! Proceeding...'
>     else:
>         quit('UNKNOWN STATE. This script may need to be updated. Submit a bug report.')
> 
>     # Now we need a very specific URL from the respose.
>     response_url = re.search("response_url\s*=\s*'(.*)';", res.get_data()).groups()[0]
>     _**agent.open(response_url)**_

response_url if I debug, looks something like this:
_https://connect.garmin.com/post-auth/login?ticket=....._

**Comment by magsol** [comment]
Ah, I see--it's this line https://github.com/magsol/garmin/blob/master/download.py#L99 

So evidently login is still succeeding, but the regular expression isn't pulling out the redirect URL correctly anymore.

**Comment by AngelaKru** [comment]
Yes, that's what seems to be happening

**Comment by AngelaKru** [comment]
I noticed that the response_url string was printing out escape characters:
_https:\/\/connect.garmin.com\/post-auth\/login?ticket=....._

I'm no Python expert, but in languages like C# there are ways you can flag a string that contains escapes so that they're not treated as just another character. I don't know if/how to do this in Python so maybe you can help there. However, I did the following which worked:

> import string


inside 'login' function:
>     #agent.open(response_url)
>    agent.open(string.replace(response_url, "\/", "/"))

**Comment by magsol** [comment]
Hmm, this is strange; I can't seem to reproduce you're error, as mine is crashing much earlier in the `login` function, at this line:

`hostname_url = agent.open(GAUTH)`

Still trying to diagnose.

**Comment by tiagovarela** [comment]
I also encountered this problem and fixed it here: https://github.com/tiagovarela/garmin/commit/cc06611028dd83cd50e1682b8a05722ae3e58d85.

To bypass the `HTTP Error 404: Not Found` I moved the `addheaders` line before the `agent.open(GAUTH)`.

I also had the problem described by @AngelaKru since the Javascript file has the forward slash encoded on the URL. After a quick search I did not find a python equivalent to the `decodeURIComponent` used so a `response_url.replace("\/", "/")` works but could probably be improved.

Hope this helps. I can create a PR with this commit if you like it.
